### ISSUE TITLE
Restarting endpoint after pg_switch_wal() fails

### ISSUE THREAD

**Issue by hlinnaka** [neondatabase/neon#8911]
## Steps to reproduce

Add this python test:

```
diff --git a/test_runner/regress/test_wal_acceptor.py b/test_runner/regress/test_wal_acceptor.py
index 3785651ae..bfb57e32c 100644
--- a/test_runner/regress/test_wal_acceptor.py
+++ b/test_runner/regress/test_wal_acceptor.py
@@ -1056,6 +1056,17 @@ def test_restart_endpoint(neon_env_builder: NeonEnvBuilder):
         random_sk.start()
         endpoint.start()
 
+# Try restarting endpoint with enabled auth.
+def test_restart_endpoint_after_switch_wal(neon_env_builder: NeonEnvBuilder):
+    env = neon_env_builder.init_start()
+
+    endpoint = env.endpoints.create_start("main")
+
+    endpoint.safe_psql("SELECT pg_switch_wal()")
+
+    endpoint.stop_and_destroy()
+    endpoint = env.endpoints.create_start("main")
+    endpoint.safe_psql("SELECT 'works'")
 
 # Context manager which logs passed time on exit.
 class DurationLogger:
```

## Expected result

Test passes.

## Actual result

Test fails:

```
------------------------------------------------------------------------------------- Captured log call --------------------------------------------------------------------------------------
2024-09-04 12:48:45.540 INFO [neon_fixtures.py:1117] Using generated control_plane_api: http://127.0.0.1:23886/upcall/v1
2024-09-04 12:48:45.546 INFO [neon_fixtures.py:1230] Config: {'default_tenant_id': '2b1be1acda13db5c0f6abae5b4a63cb2', 'broker': {'listen_addr': '127.0.0.1:23885'}, 'safekeepers': [{'id': 1, 'pg_port': 23890, 'pg_tenant_only_port': 23891, 'http_port': 23892, 'sync': False}], 'pageservers': [{'id': 1, 'listen_pg_addr': 'localhost:23888', 'listen_http_addr': 'localhost:23889', 'pg_auth_type': 'Trust', 'http_auth_type': 'Trust', 'availability_zone': 'us-east-2a', 'remote_storage': {'local_path': '/home/heikki/git-sandbox/neon/test_output/test_restart_endpoint_after_switch_wal[debug-pg16]/repo/local_fs_remote_storage/pageserver'}, 'io_buffer_alignment': None}], 'control_plane_api': 'http://127.0.0.1:23886/upcall/v1'}
2024-09-04 12:48:45.547 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local init --config=/tmp/tmpxc7_vv9a"
2024-09-04 12:48:45.561 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local storage_controller start"
2024-09-04 12:48:46.678 INFO [broker.py:34] starting storage_broker to listen incoming connections at "127.0.0.1:23885"
2024-09-04 12:48:46.680 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local pageserver start --id=1"
2024-09-04 12:48:46.680 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local safekeeper start 1 -e=--enable-offload -e=--delete-offloaded-wal -e=--partial-backup-timeout -e=10s -e=--control-file-save-interval -e=1s -e=--eviction-min-resident -e=10s"
2024-09-04 12:48:46.865 INFO [neon_fixtures.py:2663] Polling 1 for None scheduling policy and PageserverAvailability.ACTIVE availability
2024-09-04 12:48:46.875 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local tenant create --tenant-id 2b1be1acda13db5c0f6abae5b4a63cb2 --timeline-id 9c112837a4dddacab341f784bbec2e2a --pg-version 16"
2024-09-04 12:48:48.718 INFO [neon_fixtures.py:605] Initial timeline 2b1be1acda13db5c0f6abae5b4a63cb2/9c112837a4dddacab341f784bbec2e2a created successfully
2024-09-04 12:48:48.718 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local endpoint create --tenant-id 2b1be1acda13db5c0f6abae5b4a63cb2 --branch-name main --pg-version 16 --pg-port 23893 --http-port 23894 ep-1"
2024-09-04 12:48:48.723 INFO [neon_fixtures.py:4064] Starting postgres endpoint ep-1
2024-09-04 12:48:48.723 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local endpoint start --safekeepers 1 ep-1"
2024-09-04 12:48:48.876 INFO [neon_fixtures.py:4243] Postgres startup took 0.1582803726196289 seconds
2024-09-04 12:48:48.998 INFO [neon_fixtures.py:401] Executing query: SELECT pg_switch_wal()
2024-09-04 12:48:49.17  INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local endpoint stop --destroy --mode=immediate ep-1"
2024-09-04 12:48:49.227 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local endpoint create --tenant-id 2b1be1acda13db5c0f6abae5b4a63cb2 --branch-name main --pg-version 16 --pg-port 23895 --http-port 23896 ep-2"
2024-09-04 12:48:49.234 INFO [neon_fixtures.py:4064] Starting postgres endpoint ep-2
2024-09-04 12:48:49.234 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local endpoint start --safekeepers 1 ep-2"
2024-09-04 12:50:33.554 INFO [neon_fixtures.py:1679] Run ['/home/heikki/git-sandbox/neon/target/debug/neon_local', 'endpoint', 'start', '--safekeepers', '1', 'ep-2'] failed:
  stdout:
    Starting existing endpoint ep-2...
    Starting postgres node at 'postgresql://cloud_admin@127.0.0.1:23895/postgres'
    SIGKILL & wait the started process
  stderr:
    command failed: compute startup timed out; still in Init state
----------------------------------------------------------------------------------- Captured log teardown ------------------------------------------------------------------------------------
2024-09-04 12:50:33.706 INFO [neon_fixtures.py:969] Cleaning up all storage and compute nodes
2024-09-04 12:50:33.706 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local storage_controller stop"
2024-09-04 12:50:33.915 INFO [neon_fixtures.py:4472] Stopping safekeeper 1
2024-09-04 12:50:33.915 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local safekeeper stop 1"
2024-09-04 12:50:34.111 INFO [neon_fixtures.py:1937] Stopping pageserver with ['pageserver', 'stop', '--id=1']
2024-09-04 12:50:34.111 INFO [neon_fixtures.py:1613] Running command "/home/heikki/git-sandbox/neon/target/debug/neon_local pageserver stop --id=1"
================================================================================== short test summary info ===================================================================================
FAILED test_runner/regress/test_wal_acceptor.py::test_restart_endpoint_after_switch_wal[debug-pg16] - RuntimeError: Run ['/home/heikki/git-sandbox/neon/target/debug/neon_local', 'endpoint', 'start', '--safekeepers', '1', 'ep-2'] failed:
=============================================================================== 1 failed in 110.01s (0:01:50) ================================================================================
```

Compute log:

```
2024-09-04T09:45:31.479291Z  INFO build_tag: latest
2024-09-04T09:45:31.480030Z  INFO new pspec.spec: ComputeSpec { format_version: 1.0, operation_uuid: None, features: [], swap_size_bytes: None, cluster: Cluster { cluster_id: None, name: None, state: None, roles: [], databases: [], postgresql_conf: Some("max_wal_senders=10\nwal_log_hints=off\nmax_replication_slots=10\nhot_standby=on\nshared_buffers=1MB\nfsync=off\nmax_connections=100\nwal_level=logical\nwal_sender_timeout=5s\nlisten_addresses='127.0.0.1'\nport=15011\nwal_keep_size=0\nrestart_after_crash=off\nshared_preload_libraries=neon\nmax_replication_write_lag=15MB\nmax_replication_flush_lag=10GB\nsynchronous_standby_names=walproposer\nneon.safekeepers='localhost:15007'\nmax_replication_write_lag=15MB\n"), settings: None }, delta_operations: None, skip_pg_catalog_updates: true, tenant_id: Some(de0d91291a65604069eccc8a3e7bf81c), timeline_id: Some(6224c1c7133c6b8eb6734e3ff3d87470), pageserver_connstring: Some("postgresql://no_user@localhost:15004"), safekeeper_connstrings: ["127.0.0.1:15007"], mode: Primary, storage_auth_token: None, remote_extensions: None, pgbouncer_settings: None, shard_stripe_size: Some(32768) }
2024-09-04T09:45:31.482750Z  INFO running compute with features: []
2024-09-04T09:45:31.482962Z  INFO compute is not running, waiting before monitoring activity
2024-09-04T09:45:31.483055Z  INFO start_compute: starting compute for project None, operation None, tenant de0d91291a65604069eccc8a3e7bf81c, timeline 6224c1c7133c6b8eb6734e3ff3d87470
2024-09-04T09:45:31.483094Z  INFO start_compute: start_compute spec.remote_extensions None
2024-09-04T09:45:31.483121Z  INFO configurator_main_loop: waiting for reconfiguration requests
2024-09-04T09:45:31.483489Z  INFO start_compute:prepare_pgdata: checking if safekeepers are synced
2024-09-04T09:45:31.484122Z  INFO start_compute:prepare_pgdata: connecting to 127.0.0.1:15007
2024-09-04T09:45:31.484223Z  INFO starting HTTP server on [::]:15012
2024-09-04T09:45:31.493398Z  INFO start_compute:prepare_pgdata: querying 127.0.0.1:15007
2024-09-04T09:45:31.493861Z  INFO start_compute:prepare_pgdata: done with 127.0.0.1:15007
2024-09-04T09:45:31.494005Z  INFO start_compute:prepare_pgdata: not synced. 0/14F1760 0/2000000
2024-09-04T09:45:31.494081Z  INFO start_compute:prepare_pgdata: starting safekeepers syncing
PG:2024-09-04 09:45:31.497 GMT [172128] LOG:  skipping missing configuration file "/home/heikki/git-sandbox/neon/test_output/test_restart_endpoint_after_switch_wal[debug-pg16]/repo/endpoints/ep-2/pgdata/compute_ctl_temp_override.conf"
PG:2024-09-04 09:45:31.497 GMT [172128] LOG:  skipping missing configuration file "/home/heikki/git-sandbox/neon/test_output/test_restart_endpoint_after_switch_wal[debug-pg16]/repo/endpoints/ep-2/pgdata/compute_ctl_temp_override.conf"
PG:2024-09-04 09:45:31.497 GMT [172128] LOG:  skipping missing configuration file "/home/heikki/git-sandbox/neon/test_output/test_restart_endpoint_after_switch_wal[debug-pg16]/repo/endpoints/ep-2/pgdata/postgresql.auto.conf"
PG:2024-09-04 09:45:31.500 GMT [172128] LOG:  registered custom resource manager "neon" with ID 134
PG:2024-09-04 09:45:31.500 GMT [172128] LOG:  Missing NEON_CONTROL_PLANE_TOKEN environment variable, forwarding will not be authenticated
PG:2024-09-04 09:45:31.501 GMT [172128] LOG:  creating missing WAL directory "pg_wal"
PG:2024-09-04 09:45:31.501 GMT [172128] LOG:  [WP] connecting with node 127.0.0.1:15007
PG:2024-09-04 09:45:31.517 GMT [172128] LOG:  [WP] connected with node 127.0.0.1:15007
PG:2024-09-04 09:45:31.518 GMT [172128] LOG:  [WP] received AcceptorGreeting from safekeeper 127.0.0.1:15007, term=3
PG:2024-09-04 09:45:31.518 GMT [172128] LOG:  [WP] proposer connected to quorum (1) safekeepers, propTerm=4
PG:2024-09-04 09:45:31.518 GMT [172128] LOG:  [WP] requesting vote from 127.0.0.1:15007 for term 4
PG:2024-09-04 09:45:31.519 GMT [172128] LOG:  [WP] got VoteResponse from acceptor 127.0.0.1:15007, voteGiven=1, epoch=2, flushLsn=0/2000000, truncateLsn=0/14F1760, timelineStartLsn=0/14F1760
PG:2024-09-04 09:45:31.519 GMT [172128] LOG:  [WP] got votes from majority (1) of nodes, term 4, epochStartLsn 0/2000000, donor 127.0.0.1:15007, truncate_lsn 0/14F1760
PG:2024-09-04 09:45:31.519 GMT [172128] LOG:  [WP] sending elected msg to node 1 term=4, startStreamingAt=0/2000000 (lastCommonTerm=2), termHistory.n_entries=2 to 127.0.0.1:15007, timelineStartLsn=0/14F1760
PG:2024-09-04 09:45:32.038 GMT [172128] WARNING:  [WP] failed to read from node 127.0.0.1:15007 in active send state: ERROR:  WAL acceptor: truncate_wal called on non-written WAL, write_lsn=0/1511798, end_pos=0/2000000​      
PG:2024-09-04 09:45:32.501 GMT [172128] LOG:  [WP] connecting with node 127.0.0.1:15007
PG:2024-09-04 09:45:32.515 GMT [172128] LOG:  [WP] connected with node 127.0.0.1:15007
PG:2024-09-04 09:45:32.517 GMT [172128] LOG:  [WP] received AcceptorGreeting from safekeeper 127.0.0.1:15007, term=4
PG:2024-09-04 09:45:32.517 GMT [172128] LOG:  [WP] requesting vote from 127.0.0.1:15007 for term 4
PG:2024-09-04 09:45:32.517 GMT [172128] LOG:  [WP] got VoteResponse from acceptor 127.0.0.1:15007, voteGiven=0, epoch=2, flushLsn=0/2000000, truncateLsn=0/14F1760, timelineStartLsn=0/14F1760
PG:2024-09-04 09:45:32.517 GMT [172128] LOG:  [WP] sending elected msg to node 1 term=4, startStreamingAt=0/2000000 (lastCommonTerm=2), termHistory.n_entries=2 to 127.0.0.1:15007, timelineStartLsn=0/14F1760
PG:2024-09-04 09:45:32.522 GMT [172128] WARNING:  [WP] failed to read from node 127.0.0.1:15007 in active send state: ERROR:  WAL acceptor: truncate_wal called on non-written WAL, write_lsn=0/1511798, end_pos=0/2000000​      
PG:2024-09-04 09:45:33.501 GMT [172128] LOG:  [WP] connecting with node 127.0.0.1:15007
PG:2024-09-04 09:45:33.513 GMT [172128] LOG:  [WP] connected with node 127.0.0.1:15007
PG:2024-09-04 09:45:33.514 GMT [172128] LOG:  [WP] received AcceptorGreeting from safekeeper 127.0.0.1:15007, term=4
PG:2024-09-04 09:45:33.514 GMT [172128] LOG:  [WP] requesting vote from 127.0.0.1:15007 for term 4
PG:2024-09-04 09:45:33.514 GMT [172128] LOG:  [WP] got VoteResponse from acceptor 127.0.0.1:15007, voteGiven=0, epoch=2, flushLsn=0/2000000, truncateLsn=0/14F1760, timelineStartLsn=0/14F1760
PG:2024-09-04 09:45:33.514 GMT [172128] LOG:  [WP] sending elected msg to node 1 term=4, startStreamingAt=0/2000000 (lastCommonTerm=2), termHistory.n_entries=2 to 127.0.0.1:15007, timelineStartLsn=0/14F1760
PG:2024-09-04 09:45:33.518 GMT [172128] WARNING:  [WP] failed to read from node 127.0.0.1:15007 in active send state: ERROR:  WAL acceptor: truncate_wal called on non-written WAL, write_lsn=0/1511798, end_pos=0/2000000​      
...
```

**Comment by hlinnaka** [comment]
If the compute writes any WAL record after the pg_switch_wal(), for example with "select pg_current_xact_id()", then the test passes.

**Comment by jcsp** [comment]
@arssher can you speculate what might be going wrong here?

**Comment by arssher** [comment]
https://github.com/neondatabase/neon/pull/8943

**Comment by hlinnaka** [comment]
Fixed by PR #8943
